File: Tables.cs
namespace SNI
{
    using System;
    using ServiceStack.DataAnnotations;

    public enum Role
    {
        Admin, // 管理员
        Reader // 读者
    }

    public enum Frequency
    {
        Annually,     // 年刊
        Semiannually, // 半年刊
        Quarterly,    // 季刊
        Monthly,      // 月刊
        Semimonthly,  // 半月刊
        Fortnightly,  // 旬刊
        Weekly        // 周刊
    }

    public class User
    {
        [PrimaryKey] [AutoIncrement] public int UserId { get; set; }
        [Required] [Unique] public Guid SessionId { get; set; }
        [Required] [Unique] public string Username { get; set; } // 用户名
        [Required] public string Nickname { get; set; }          // 昵称
        [Required] public string Password { get; set; }          // 密码
        [Required] public Role Role { get; set; }                // 角色
        [Required] public string FirstName { get; set; }         // 名字
        [Required] public string LastName { get; set; }          // 姓氏
        [Required] public string EmailAddress { get; set; }      // 邮箱
        [Required] public string PhoneNumber { get; set; }       // 手机号
    }

    public class Journal
    {
        [PrimaryKey] [AutoIncrement] public int JournalId { get; set; }
        [Required] public string Name { get; set; }          // 期刊名称
        [Required] public string UsedName { get; set; }      // 曾用名
        [Required] public string Language { get; set; }      // 语种
        [Required] public Frequency Frequency { get; set; }  // 出版周期
        [Required] [Unique] public string Issn { get; set; } // ISSN 刊号
        [Required] [Unique] public string Cn { get; set; }   // CN 刊号
        [Required] [Unique] public string Pdc { get; set; }  // 邮发代号
        [Required] public string History { get; set; }       // 创刊时间
        [Required] public string Organizer { get; set; }     // 主办单位
        [Required] public string Publisher { get; set; }     // 出版单位
        [Required] public string Address { get; set; }       // 汇款地址
    }

    public class Subscription
    {
        [PrimaryKey] [AutoIncrement] public int SubscriptionId { get; set; }
        [Required] public int JournalId { get; set; } // 期刊 ID
        [Required] public int Year { get; set; }      // 征订年份
    }

    public class Storage
    {
        [PrimaryKey] [AutoIncrement] public int StorageId { get; set; }
        [Required] public int JournalId { get; set; } // 期刊 ID
        [Required] public int Year { get; set; }      // 年
        [Required] public int Volume { get; set; }    // 卷
        [Required] public int Issue { get; set; }     // 期
    }

    public class Article
    {
        [PrimaryKey] [AutoIncrement] public int ArticleId { get; set; }
        [Required] public int StorageId { get; set; }   // 库存 ID
        [Required] public int PageNumber { get; set; }  // 页码
        [Required] public string Title { get; set; }    // 标题
        [Required] public string Author { get; set; }   // 作者
        [Required] public string Content { get; set; }  // 内容
        [Required] public string Keyword1 { get; set; } // 关键字
        [Required] public string Keyword2 { get; set; }
        [Required] public string Keyword3 { get; set; }
        [Required] public string Keyword4 { get; set; }
        [Required] public string Keyword5 { get; set; }
    }

    public class Borrowing
    {
        [PrimaryKey] [AutoIncrement] public int BorrowingId { get; set; }
        [Required] public int UserId { get; set; }          // 用户 ID
        [Required] public int StorageId { get; set; }       // 库存 ID
        [Required] public DateTime BorrowTime { get; set; } // 借出时间
        [Required] public DateTime AgreedTime { get; set; } // 应还时间
        public DateTime ReturnTime { get; set; }            // 归还时间
    }
}

File: tables.py
#!/usr/bin/env/python3
# -*- coding: utf-8 -*-
from datetime import date, datetime

from pony import orm


class EntityMeta(orm.core.EntityMeta):
    @staticmethod
    def clean_kwargs(kwargs):
        return {k: kwargs[k] for k in kwargs
                if kwargs[k] is not None}

    @orm.db_session
    def insert(cls, *args, **kwargs):
        kwargs = cls.clean_kwargs(kwargs)
        return cls(*args, **kwargs)

    @orm.db_session
    def exists(cls, *args, **kwargs):
        kwargs = cls.clean_kwargs(kwargs)
        return super().exists(*args, **kwargs)

    @orm.db_session
    def get(cls, *args, **kwargs):
        kwargs = cls.clean_kwargs(kwargs)
        return super().get(*args, **kwargs)

    @orm.db_session
    def select(cls, *args, **kwargs):
        if kwargs:
            kwargs = cls.clean_kwargs(kwargs)
            return super().select().filter(**kwargs)[:]
        return super().select(*args)[:]


@orm.db_session
def delete_db(self):
    db.Entity.delete_db(self)


@orm.db_session
def set_db(self, **kwargs):
    db.Entity.set_db(self, **kwargs)


db = orm.Database()
db.Entity.delete_db = db.Entity.delete
db.Entity.set_db = db.Entity.set
db.Entity.delete = delete_db
db.Entity.set = set_db


class User(db.Entity, metaclass=EntityMeta):
    user_id = orm.PrimaryKey(int, auto=True)
    username = orm.Required(str, unique=True)
    nickname = orm.Required(str)
    password = orm.Required(str)
    first_name = orm.Optional(str)
    last_name = orm.Optional(str)
    email_address = orm.Optional(str)
    phone_number = orm.Optional(str)
    session_id = orm.Required(str)
    expire_time = orm.Required(datetime)


class Administrator(User): pass


class Reader(User): pass


class Journal(db.Entity, metaclass=EntityMeta):
    journal_id = orm.PrimaryKey(int, auto=True)
    name = orm.Required(str)
    language = orm.Required(str)
    history = orm.Required(str)
    host = orm.Required(str)
    publisher = orm.Required(str)
    used_name = orm.Optional(str)
    frequency = orm.Required(str)
    issn = orm.Required(str, unique=True)
    cn = orm.Required(str, unique=True)
    pdc = orm.Required(str, unique=True)


class Subscription(db.Entity, metaclass=EntityMeta):
    subscription_id = orm.PrimaryKey(int, auto=True)
    journal_id = orm.Required(int)
    year = orm.Required(int)
    orm.composite_key(journal_id, year)


class Storage(db.Entity, metaclass=EntityMeta):
    storage_id = orm.PrimaryKey(int, auto=True)
    journal_id = orm.Required(int)
    year = orm.Required(int)
    volume = orm.Required(int)
    issue = orm.Required(int)
    orm.composite_key(journal_id, year, volume, issue)


class Article(db.Entity, metaclass=EntityMeta):
    article_id = orm.PrimaryKey(int, auto=True)
    storage_id = orm.Required(int)
    page_number = orm.Required(int)
    title = orm.Required(str)
    author = orm.Required(str)
    content = orm.Required(str)
    keyword1 = orm.Optional(str)
    keyword2 = orm.Optional(str)
    keyword3 = orm.Optional(str)
    keyword4 = orm.Optional(str)
    keyword5 = orm.Optional(str)


class Borrowing(db.Entity, metaclass=EntityMeta):
    borrowing_id = orm.PrimaryKey(int, auto=True)
    user_id = orm.Required(int)
    storage_id = orm.Required(int)
    borrow_date = orm.Required(date)
    agreed_date = orm.Required(date)
    return_date = orm.Optional(date)


def bind_sqlite(filename=':memory:'):
    db.bind('sqlite', filename, create_db=True)
    db.generate_mapping(create_tables=True)
    with orm.db_session:
        # db.execute('PRAGMA synchronous = OFF')
        db.execute('PRAGMA journal_mode = WAL')
